### PR TITLE
Backports 0.15 pr13

### PR DESCRIPTION
**Pull request by PastaPastaPasta**

### CODE REVIEW

**Comment by UdjinM6** [comment]
need rebase

**Comment by PastaPastaPasta** [comment]
Rebased, couple of conflicts resolved

File: src/qt/clientmodel.cpp [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #include "guiutil.h"
 #include "peertablemodel.h"
 
+#include "chain.h"
```
Comment (UdjinM6) [inline src/qt/clientmodel.cpp]: should be one line lower

File: test/functional/README.md [hunk 2]
```diff
@@ -59,7 +59,7 @@ thread.)
 * RPC calls are available in p2p tests.
 
 * Can be used to write free-form tests, where specific p2p-protocol behavior
-is tested.  Examples: ```p2p-accept-block.py```, ```maxblocksinflight.py```.
+is tested.  Examples: ```p2p-accept-block.py```, ```p2p-compactblocks.py```.
```
Comment (UdjinM6) [inline test/functional/README.md]: should also drop `maxblocksinflight.py` file itself

File: test/functional/zmq_test.py [hunk 3]
```diff
@@ -18,6 +20,21 @@ def __init__(self):
...
+        except ImportError:
+            self.log.warning("python3-zmq module not available. Skipping zmq tests!")
+            sys.exit(self.TEST_EXIT_SKIPPED)
+
+        # Check that bitcoin has been built with ZMQ enabled
+        config = configparser.ConfigParser()
+        config.read_file(open(os.path.dirname(__file__) + "/config.ini"))
+
+        if not config["components"].getboolean("ENABLE_ZMQ"):
+            self.log.warning("bitcoind has not been built with zmq enabled. Skipping zmq tests!")
```
Comment (UdjinM6) [inline test/functional/zmq_test.py]: s/bitcoind/dashd/

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -4191,7 +4157,7 @@ void CWallet::KeepKey(int64_t nIndex)
     // Remove from key pool
     if (fFileBacked)
     {
-        CWalletDB walletdb(strWalletFile);
+        CWalletDB walletdb(*dbw);
```
Comment (UdjinM6) [inline src/wallet/wallet.cpp]: should drop ` if (fFileBacked)` wrapper above

**Comment by PastaPastaPasta** [comment]
resolved comments, redid 10201, there's just a lot of segwit stuff built in that seemed to have been changed

File: src/validation.cpp [hunk 5]
```diff
@@ -632,6 +635,12 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
     if (tx.IsCoinBase())
         return state.DoS(100, false, REJECT_INVALID, "coinbase");
 
+    // Reject transactions with witness before segregated witness activates (override with -prematurewitness)
```
Comment (UdjinM6) [inline src/validation.cpp]: this should be dropped

File: src/validation.cpp [hunk 6]
```diff
@@ -883,20 +892,21 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
     return true;
 }
 
-bool AcceptToMemoryPoolWithTime(CTxMemPool& pool, CValidationState &state, const CTransactionRef &tx, bool fLimitFree,
+/** (try to) add transaction to memory pool with a specified acceptance time **/
+static bool AcceptToMemoryPoolWithTime(const CChainParams& chainparams, CTxMemPool& pool, CValidationState &state, const CTransactionRef &tx, bool fLimitFree,
```
Comment (UdjinM6) [inline src/validation.cpp]: missing changes in `AcceptToMemoryPool`

File: src/validation.cpp [hunk 7]
```diff
@@ -3359,7 +3375,7 @@ static bool CheckIndexAgainstCheckpoint(const CBlockIndex* pindexPrev, CValidati
     return true;
 }
 
-bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, const CBlockIndex* pindexPrev, int64_t nAdjustedTime)
+static bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationState& state, const Consensus::Params& consensusParams, const CBlockIndex* pindexPrev, int64_t nAdjustedTime)
```
Comment (UdjinM6) [inline src/validation.cpp]: missing comments from original PR

**Comment by PastaPastaPasta** [comment]
addresses comments, can squash after review :)

**Comment by PastaPastaPasta** [comment]
tried to include 8524, still not building. Will work more on it later

**Comment by PastaPastaPasta** [comment]
removed 10201 and 8524, they will be in a second pr.
builds and passes tests

**Comment by UdjinM6** [comment]
LGTM. Would be nice to have fixes squashed into original "Merge bitcoin#xxxx"  commits now.

**Comment by PastaPastaPasta** [comment]
squashed

**Comment by PastaPastaPasta** [comment]
@codablock please review, as you would say LGTM ;)

**Pull request merged**